File: java/src/main/java/io/binac/leetcode/ConstructBinaryTreeFromPreorderAndInorderTraversal.java
package io.binac.leetcode;

import io.binac.leetcode.util.TreeNode;

/**
 * <p>Given two integer arrays <code>preorder</code> and <code>inorder</code> where <code>preorder</code> is the preorder traversal of a binary tree and <code>inorder</code> is the inorder traversal of the same tree, construct and return <em>the binary tree</em>.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 * <img alt="" src="https://assets.leetcode.com/uploads/2021/02/19/tree.jpg" style="width: 277px; height: 302px;">
 * <pre><strong>Input:</strong> preorder = [3,9,20,15,7], inorder = [9,3,15,20,7]
 * <strong>Output:</strong> [3,9,20,null,null,15,7]
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> preorder = [-1], inorder = [-1]
 * <strong>Output:</strong> [-1]
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>1 &lt;= preorder.length &lt;= 3000</code></li>
 * 	<li><code>inorder.length == preorder.length</code></li>
 * 	<li><code>-3000 &lt;= preorder[i], inorder[i] &lt;= 3000</code></li>
 * 	<li><code>preorder</code> and <code>inorder</code> consist of <strong>unique</strong> values.</li>
 * 	<li>Each value of <code>inorder</code> also appears in <code>preorder</code>.</li>
 * 	<li><code>preorder</code> is <strong>guaranteed</strong> to be the preorder traversal of the tree.</li>
 * 	<li><code>inorder</code> is <strong>guaranteed</strong> to be the inorder traversal of the tree.</li>
 * </ul>
 */
public class ConstructBinaryTreeFromPreorderAndInorderTraversal {
    public static class Solution1 {
        private TreeNode buildTree(int[] preOrder, int prePos, int[] inOrder, int inPos, int len) {
            if (len == 0)
                return null;

            int pos = inPos;
            for (int e = inPos + len; pos < e && inOrder[pos] != preOrder[prePos]; ++pos) ;
            final int left = pos - inPos, right = len - left - 1;
            return new TreeNode(preOrder[prePos], buildTree(preOrder, prePos + 1, inOrder, inPos, left), buildTree(preOrder, prePos + 1 + left, inOrder, pos + 1, right));
        }

        public TreeNode buildTree(int[] preorder, int[] inorder) {
            return buildTree(preorder, 0, inorder, 0, preorder.length);
        }
    }
}

File: python3/leetcodepy/construct_binary_tree_from_preorder_and_inorder_traversal.py
"""
Given two integer arrays preorder and inorder where preorder is the preorder traversal of a binary tree and inorder is the inorder traversal of the same tree, construct and return the binary tree.



Example 1:
https://assets.leetcode.com/uploads/2021/02/19/tree.jpg

Input: preorder = [3,9,20,15,7], inorder = [9,3,15,20,7]
Output: [3,9,20,null,null,15,7]

Example 2:

Input: preorder = [-1], inorder = [-1]
Output: [-1]


Constraints:

1 <= preorder.length <= 3000
inorder.length == preorder.length
-3000 <= preorder[i], inorder[i] <= 3000
preorder and inorder consist of unique values.
Each value of inorder also appears in preorder.
preorder is guaranteed to be the preorder traversal of the tree.
inorder is guaranteed to be the inorder traversal of the tree.
"""
from typing import List, Optional

from .utils import TreeNode


class Solution1:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder:
            return None
        i = inorder.index(preorder[0])
        return TreeNode(preorder[0], self.buildTree(preorder[1:i + 1], inorder[:i]),
                        self.buildTree(preorder[i + 1:], inorder[i + 1:]))
